File: CGConvertor/UpdateChecker.swift
import AppKit

/// "Check for Updates" — compară versiunea rulată cu ultimul tag de pe
/// GitHub Releases și descarcă+instalează automat noua versiune, FĂRĂ să
/// mai treacă prin browser/pagina de GitHub — vezi SelfUpdater.swift și
/// CLAUDE.md Partea 1, Regula 20.
enum UpdateChecker {
    private static let latestReleaseAPIURL = URL(string: "https://api.github.com/repos/gordasgdc/CGConvertor/releases/latest")!
    private static let releasesPageURL = URL(string: "https://github.com/gordasgdc/CGConvertor/releases/latest")!

    static var currentVersion: String {
        Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "?"
    }

    /// Verificare automată, o singură dată per lansare, tăcută dacă nu e
    /// nimic nou. Dismissal per-versiune, ca update-ul deja văzut să nu
    /// reapară la fiecare pornire.
    static func checkSilentlyOnLaunch(onNewVersion: @escaping (String, URL) -> Void) {
        Task {
            if case .newVersion(let version, let pkgURL) = await fetchLatestTag() {
                let dismissedKey = "cgconvertor_dismissed_update_version"
                if UserDefaults.standard.string(forKey: dismissedKey) == version { return }
                await MainActor.run { onNewVersion(version, pkgURL) }
            }
        }
    }

    static func markDismissed(_ version: String) {
        UserDefaults.standard.set(version, forKey: "cgconvertor_dismissed_update_version")
    }

    static func checkAndShowAlert() {
        Task {
            let result = await fetchLatestTag()
            await MainActor.run { presentResult(result) }
        }
    }

    private enum Result {
        case upToDate
        case newVersion(String, URL)
        case error
    }

    private static func fetchLatestTag() async -> Result {
        var request = URLRequest(url: latestReleaseAPIURL)
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let tag = json["tag_name"] as? String else {
                return .error
            }
            let latest = tag.hasPrefix("v") ? String(tag.dropFirst()) : tag
            guard isVersion(latest, newerThan: currentVersion) else { return .upToDate }

            // Asset-ul .pkg cu nume STABIL ("CGConvertor.pkg") publicat de
            // build_installer.sh la fiecare release.
            let assets = json["assets"] as? [[String: Any]] ?? []
            let pkgAsset = assets.first { ($0["name"] as? String) == "CGConvertor.pkg" }
            guard let urlString = pkgAsset?["browser_download_url"] as? String, let pkgURL = URL(string: urlString) else {
                return .newVersion(latest, releasesPageURL)
            }
            return .newVersion(latest, pkgURL)
        } catch {
            return .error
        }
    }

    private static func isVersion(_ a: String, newerThan b: String) -> Bool {
        let partsA = a.split(separator: ".").compactMap { Int($0) }
        let partsB = b.split(separator: ".").compactMap { Int($0) }
        for i in 0..<max(partsA.count, partsB.count) {
            let x = i < partsA.count ? partsA[i] : 0
            let y = i < partsB.count ? partsB[i] : 0
            if x != y { return x > y }
        }
        return false
    }

    private static func presentResult(_ result: Result) {
        let alert = NSAlert()
        switch result {
        case .upToDate:
            alert.messageText = L.t("update.upToDate.title")
            alert.informativeText = String(format: L.t("update.upToDate.body"), currentVersion)
            alert.addButton(withTitle: "OK")
            alert.runModal()
        case .newVersion(let version, let pkgURL):
            alert.messageText = L.t("update.available.title")
            alert.informativeText = String(format: L.t("update.available.body"), version, currentVersion)
            alert.addButton(withTitle: L.t("update.download"))
            alert.addButton(withTitle: L.t("update.later"))
            let response = alert.runModal()
            markDismissed(version)
            if response == .alertFirstButtonReturn {
                Task { await SelfUpdater.downloadAndInstall(pkgURL: pkgURL, version: version) }
            }
        case .error:
            alert.messageText = L.t("update.error.title")
            alert.informativeText = L.t("update.error.body")
            alert.addButton(withTitle: "OK")
            alert.runModal()
        }
    }
}

File: python/update_checker.py
"""
update_checker.py — verificare manuala/automata de versiune noua pentru
CG Convertor, citind ultimul tag de pe GitHub Releases (acelasi API pe
care il foloseste varianta Swift, UpdateChecker.swift — nu update.json,
CG Convertor nu are unul separat). Nu e updater silentios/automat — doar
anunta si ofera un link de descarcare, la fel ca restul ecosistemului GDC.
"""

import json
import re
import urllib.request
import urllib.error

LATEST_RELEASE_API_URL = "https://api.github.com/repos/gordasgdc/CGConvertor/releases/latest"
RELEASES_PAGE_URL = "https://github.com/gordasgdc/CGConvertor/releases/latest"


def _version_tuple(version_string):
    parts = []
    for piece in re.split(r"[.\-]", version_string.strip().lstrip("vV")):
        match = re.match(r"\d+", piece)
        parts.append(int(match.group()) if match else 0)
    return tuple(parts) if parts else (0,)


def is_newer_version(candidate, current):
    return _version_tuple(candidate) > _version_tuple(current)


def check_for_updates(current_version, timeout=10):
    """Intoarce {"available": bool, "version": str|None, "download_url": str|None, "error": str|None}.

    `download_url` e link-ul DIRECT al asset-ului installer-ului
    (`CGConvertor-Windows-Setup.exe`, nume stabil publicat de CI la
    fiecare release) - folosit de self_updater.py, NU de un browser.
    Vezi CLAUDE.md Partea 1, Regula 20.
    """
    try:
        request = urllib.request.Request(
            LATEST_RELEASE_API_URL,
            headers={"User-Agent": "CGConvertor-Updater", "Accept": "application/vnd.github+json"},
        )
        with urllib.request.urlopen(request, timeout=timeout) as response:
            data = json.loads(response.read().decode("utf-8"))
        tag = str(data.get("tag_name", "")).strip()
        if not tag:
            return {"available": False, "error": "Raspuns invalid de la GitHub."}
        latest = tag[1:] if tag.lower().startswith("v") else tag
        if is_newer_version(latest, current_version):
            download_url = None
            for asset in data.get("assets", []):
                if asset.get("name") == "CGConvertor-Windows-Setup.exe":
                    download_url = asset.get("browser_download_url")
                    break
            return {"available": True, "version": latest, "download_url": download_url, "error": None}
        return {"available": False, "error": None}
    except urllib.error.URLError as e:
        return {"available": False, "error": f"Nu am putut contacta GitHub: {e.reason}"}
    except Exception as e:
        return {"available": False, "error": str(e)}
